packages/docprovider/src/notebookCellExecutor.ts

```diff
@@ -0,0 +1,310 @@
+      case 'code':
+        if (sessionContext) {
+          if (sessionContext.isTerminating) {
+            await showDialog({
+              title: trans.__('Kernel Terminating'),
+              body: trans.__(
+                'The kernel for %1 appears to be terminating. You can not run any cell for now.',
+                sessionContext.session?.path
+              ),
+              buttons: [Dialog.okButton()]
+            });
+            break;
+          }
+          if (sessionContext.pendingInput) {
+            await showDialog({
+              title: trans.__('Cell not executed due to pending input'),
+              body: trans.__(
+                'The cell has not been executed to avoid kernel deadlock as there is another pending input! Submit your pending input and try again.'
+              ),
+              buttons: [Dialog.okButton()]
+            });
+            return false;
+          }
+          if (sessionContext.hasNoKernel) {
+            const shouldSelect = await sessionContext.startKernel();
+            if (shouldSelect && sessionDialogs) {
+              await sessionDialogs.selectKernel(sessionContext);
+            }
+          }
+
+          if (sessionContext.hasNoKernel) {
+            cell.model.sharedModel.transact(() => {
+              (cell.model as ICodeCellModel).clearExecution();
+            });
+            return true;
+          }
+
+          const kernelId = sessionContext?.session?.kernel?.id;
+          const apiURL = URLExt.join(
+            this._serverSettings.baseUrl,
+            `api/kernels/${kernelId}/execute`
+          );
+          const cellId = cell.model.sharedModel.getId();
+          const documentId = notebook.sharedModel.getState('room_id');
+
+          const init = {
+            method: 'POST',
+            body: JSON.stringify({ cell_id: cellId, document_id: documentId })
+          };
+          onCellExecutionScheduled({ cell });
+          let success = false;
+          try {
+            // FIXME quid of deletedCells and timing record
+            const response = await requestServer(
+              cell as CodeCell,
+              apiURL,
+              init,
+              this._serverSettings,
+              translator
+            );
+            const data = await response.json();
+            success = data['status'] === 'ok';
+          } catch (error: any) {
+            onCellExecuted({ cell, success: false, error });
+            if (cell.isDisposed) {
+              return false;
+            } else {
+              throw await ServerConnection.ResponseError.create(error);
+            }
+          }
+
+          onCellExecuted({ cell, success });
+
+          return true;
+        }
+        cell.model.sharedModel.transact(() => {
+          (cell.model as ICodeCellModel).clearExecution();
+        }, false);
+        break;
+      default:
+        break;
+    }
+    return Promise.resolve(true);
+  }
+}
+
+async function requestServer(
```

Comment:
I'm not sure about this logic. I understand it's a way to have input requests and widgets to work, but the whole system is designed so that it works with https://github.com/datalayer/jupyter-server-nbmodel, which is a third-party extension.
There are other ways to support inputs and widgets, discussed in https://github.com/jupyter-server/jupyter_ydoc/pull/227#issuecomment-2049778352. They involve using CRDTs so that they are fully integrated into `jupyter-collaboration`.

Reply:
> That part is definitely opinionated. It only highlights the need to define a standard way to deal with input. I have no problem changing that part. A possibility to allow testing of various approach could be to remove the modification in the cell executor that are specific to `jupyter-server-nbmodel`. That extension will then provide its own executor independently of collaboration.

I think it would make sense indeed to define an API for handling inputs, as you did for cell execution. Correct me if I'm wrong, but the way it is handled in `jupyter-server-nbmodel` doesn't support state recovery? For instance if I run a cell with `a = input()`, close my browser, and reopen it, I don't have the input widget anymore and my notebook is stuck (I can't run a cell anymore). For me this highlights the need for a better solution based on CRDTs.
Having an API for input handling would allow to provide a CRDT-based input widget.